libqtile/backend/x11/xcbq.py

```diff
@@ -802,6 +803,52 @@ def grab_pointer(self, owner_events, event_mask, pointer_mode,
     def ungrab_pointer(self):
         self.conn.conn.core.UngrabPointer(xcffib.xproto.Atom._None)
 
+    def paint_borders(self, inner_w, inner_h, borderwidths, bordercolors):
+        if len(bordercolors) == 1 or len(bordercolors) != len(borderwidths):
+            self.set_attribute(borderpixel=bordercolors[0])
+            return
+
+        core = self.conn.conn.core
+        self.borderwidth = sum(borderwidths)
+        outer_w = inner_w + self.borderwidth * 2
+        outer_h = inner_h + self.borderwidth * 2
+
+        pixmap = self.conn.conn.generate_id()
+        core.CreatePixmap(
+            self.conn.default_screen.root_depth, pixmap, self.wid, outer_w, outer_h
+        )
+        gc = self.conn.conn.generate_id()
+        core.CreateGC(gc, pixmap, 0, None)
+
+        coord = 0
+        for i in range(len(borderwidths)):
+            core.ChangeGC(gc, xcffib.xproto.GC.Foreground, [bordercolors[i]])
+            rect = xcffib.xproto.RECTANGLE.synthetic(
+                coord, coord, outer_w - coord * 2, outer_h - coord * 2
+            )
+            core.PolyFillRectangle(pixmap, gc, 1, [rect])
+            coord += borderwidths[i]
+
+        self.set_borderpixmap(pixmap, gc, outer_w, outer_h)
+        core.FreePixmap(pixmap)
```

Comment:
These will all leak if something in the middle throws an exception; can we wrap this in some kind of try: finally or context handler or something?